File: src/base_files/nodejs/core/utils/typedef.ts
import { Failure } from "../errors/failures";

type Result<Type> = Failure | Type;
export type ResultFuture<Type> = Promise<Result<Type>>;
export type ResultFutureVoid = ResultFuture<void>;

export type DataMap = {
    [key: string]: any;
};

export type ApiResponse<Type> = {
    data?: Type,
    total?: number,
    pages?: number,
    page?: number,
    perPage?: number
    message?: string
};
export interface ApiParams {
    limit?: number,
    page?: number,
    offset?: number,
    perPage?: number,
    search?: string,
}

File: src/base_files/dart/core/utils/typedef.dart
import 'package:dartz/dartz.dart';

import '../errors/failures.dart';

typedef Result<Type> = Either<Failure, Type>;
typedef ResultFuture<Type> = Future<Result<Type>>;
typedef ResultFutureVoid = ResultFuture<void>;

typedef DataMap = Map<String, dynamic>;

class RemoteServiceResponse<Type> {
  Type data;
  int? total;
  int? pages;
  int? page;
  int? perPage;
  String? message;

  RemoteServiceResponse(
      {required this.data,
      this.total,
      this.pages,
      this.page,
      this.perPage,
      this.message});

  RemoteServiceResponse.fromJson(DataMap json, Type data)
      : this(
          data: data,
          total: json["total"],
          pages: json["pages"],
          page: json["page"],
          perPage: json["perPage"],
          message: json["message"],
        );

  DataMap toJson() => {
        "data": data,
        "total": total,
        "pages": pages,
        "page": page,
        "perPage": perPage,
        "message": message,
      };
}

class RemoteServiceParams {
  int? limit;
  int? page;
  int? offset;
  int? perPage;
  String? search;

  RemoteServiceParams(
      {this.limit, this.page, this.offset, this.perPage, this.search});

  DataMap toJson() => {
        "limit": limit,
        "page": page,
        "offset": offset,
        "perPage": perPage,
        "search": search
      };
}
